File: core/runtime.py
"""运行时判定与目录约定。"""

from __future__ import annotations

import os
import sys
from dataclasses import dataclass
from enum import Enum
from pathlib import Path


APP_DIR_NAME = "ATRI-Chat"
PORTABLE_MARKER = "portable.mode"
BINARIES_DIR_NAME = "binaries"
RESOURCES_DIR_NAME = "resources"


class AppEnv(str, Enum):
    """应用环境。"""

    DEVELOPMENT = "development"
    PRODUCTION = "production"


class RuntimeMode(str, Enum):
    """桌面端运行模式。"""

    DEVELOPMENT = "development"
    INSTALLED = "installed"
    PORTABLE = "portable"


@dataclass(frozen=True)
class RuntimeLayout:
    """运行时判定结果。"""

    app_env: AppEnv
    mode: RuntimeMode
    app_root: Path
    data_root: Path
    logs_root: Path


def get_project_root() -> Path:
    """返回仓库根目录。"""
    return Path(__file__).resolve().parent.parent


def get_local_data_home() -> Path:
    """返回本地用户数据根目录。"""
    local_appdata = os.environ.get("LOCALAPPDATA")
    if local_appdata:
        return Path(local_appdata).resolve() / APP_DIR_NAME
    return get_project_root()


def resolve_data_root(runtime_mode: RuntimeMode, app_root: Path) -> Path:
    """根据运行模式解析数据根目录。"""
    if runtime_mode == RuntimeMode.INSTALLED:
        return get_local_data_home() / "data"
    return app_root / "data"


def resolve_logs_root(runtime_mode: RuntimeMode, app_root: Path) -> Path:
    """根据运行模式解析日志根目录。"""
    if runtime_mode == RuntimeMode.INSTALLED:
        return get_local_data_home() / "logs"
    return app_root / "data" / "logs"


def resolve_runtime_layout() -> RuntimeLayout:
    """根据实际运行情况判定目录策略。"""
    if not getattr(sys, "frozen", False):
        app_root = get_project_root().resolve()
        mode = RuntimeMode.DEVELOPMENT
        return RuntimeLayout(
            app_env=AppEnv.DEVELOPMENT,
            mode=mode,
            app_root=app_root,
            data_root=resolve_data_root(mode, app_root).resolve(),
            logs_root=resolve_logs_root(mode, app_root).resolve(),
        )

    executable = Path(sys.executable).resolve()
    app_root = resolve_packaged_app_root(executable)
    mode = detect_packaged_mode(app_root)
    return RuntimeLayout(
        app_env=AppEnv.PRODUCTION,
        mode=mode,
        app_root=app_root,
        data_root=resolve_data_root(mode, app_root).resolve(),
        logs_root=resolve_logs_root(mode, app_root).resolve(),
    )


def resolve_packaged_app_root(executable: Path) -> Path:
    """根据打包产物位置反推出应用根目录。"""
    exe_dir = executable.parent
    if exe_dir.name.lower() != BINARIES_DIR_NAME:
        return exe_dir

    parent = exe_dir.parent
    if parent.name.lower() == RESOURCES_DIR_NAME and parent.parent:
        return parent.parent
    return parent


def detect_packaged_mode(app_root: Path) -> RuntimeMode:
    """根据打包目录结构判断安装版或便携版。"""
    if (app_root / PORTABLE_MARKER).exists():
        return RuntimeMode.PORTABLE
    return RuntimeMode.INSTALLED

File: frontend/src-tauri/src/runtime.rs
use std::env;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use tauri::Manager;

const APP_DIR_NAME: &str = "ATRI-Chat";
const PORTABLE_MARKER: &str = "portable.mode";
const BINARIES_DIR_NAME: &str = "binaries";
const RESOURCES_DIR_NAME: &str = "resources";

#[derive(Clone, Copy, Debug)]
pub enum RuntimeMode {
    Development,
    Installed,
    Portable,
}

impl RuntimeMode {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Development => "development",
            Self::Installed => "installed",
            Self::Portable => "portable",
        }
    }
}

#[derive(Clone, Debug)]
pub struct RuntimeLayout {
    pub mode: RuntimeMode,
    pub app_root: PathBuf,
    pub data_root: PathBuf,
    pub logs_root: PathBuf,
}

impl RuntimeLayout {
    pub fn ensure_directories(&self) -> io::Result<()> {
        fs::create_dir_all(&self.data_root)?;
        fs::create_dir_all(&self.logs_root)?;
        Ok(())
    }
}

pub fn resolve_runtime_layout() -> RuntimeLayout {
    if cfg!(debug_assertions) {
        let app_root = development_app_root();
        return RuntimeLayout {
            mode: RuntimeMode::Development,
            data_root: app_root.join("data"),
            logs_root: app_root.join("data").join("logs"),
            app_root,
        };
    }

    let app_root = packaged_app_root();
    let mode = detect_packaged_mode(&app_root);
    let local_data_root = local_appdata_root()
        .unwrap_or_else(|| app_root.clone())
        .join(APP_DIR_NAME);

    match mode {
        RuntimeMode::Installed => RuntimeLayout {
            mode,
            data_root: local_data_root.join("data"),
            logs_root: local_data_root.join("logs"),
            app_root,
        },
        RuntimeMode::Portable => RuntimeLayout {
            mode,
            data_root: app_root.join("data"),
            logs_root: app_root.join("data").join("logs"),
            app_root,
        },
        RuntimeMode::Development => RuntimeLayout {
            mode,
            data_root: app_root.join("data"),
            logs_root: app_root.join("data").join("logs"),
            app_root,
        },
    }
}

pub fn resolve_backend_path(app: &tauri::AppHandle) -> Option<PathBuf> {
    let backend_name = backend_binary_name();

    if cfg!(debug_assertions) {
        return Some(PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("binaries").join(backend_name));
    }

    let resource_dir = app.path().resource_dir().ok()?;
    Some(resource_dir.join("binaries").join(backend_name))
}

fn detect_packaged_mode(app_root: &Path) -> RuntimeMode {
    if app_root.join(PORTABLE_MARKER).exists() {
        RuntimeMode::Portable
    } else {
        RuntimeMode::Installed
    }
}

fn development_app_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .canonicalize()
        .unwrap_or_else(|_| PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("..").join(".."))
}

fn packaged_app_root() -> PathBuf {
    env::current_exe()
        .ok()
        .map(|path| resolve_app_root_from_executable(&path))
        .unwrap_or_else(|| env::current_dir().unwrap_or_else(|_| PathBuf::from(".")))
}

fn resolve_app_root_from_executable(executable: &Path) -> PathBuf {
    let exe_dir = executable
        .parent()
        .map(|dir| dir.to_path_buf())
        .unwrap_or_else(|| PathBuf::from("."));

    if exe_dir.file_name().and_then(|name| name.to_str()) != Some(BINARIES_DIR_NAME) {
        return exe_dir;
    }

    let parent = exe_dir.parent().map(|dir| dir.to_path_buf()).unwrap_or(exe_dir);
    if parent.file_name().and_then(|name| name.to_str()) == Some(RESOURCES_DIR_NAME) {
        return parent.parent().map(|dir| dir.to_path_buf()).unwrap_or(parent);
    }

    parent
}

fn local_appdata_root() -> Option<PathBuf> {
    env::var_os("LOCALAPPDATA").map(PathBuf::from)
}

fn backend_binary_name() -> &'static str {
    if cfg!(windows) {
        "atri-backend-x86_64-pc-windows-msvc.exe"
    } else if cfg!(target_os = "macos") {
        if cfg!(target_arch = "aarch64") {
            "atri-backend-aarch64-apple-darwin"
        } else {
            "atri-backend-x86_64-apple-darwin"
        }
    } else {
        "atri-backend-x86_64-unknown-linux-gnu"
    }
}
